federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+		fedHpa.Status.DesiredReplicas != updatedStatus.DesiredReplicas ||
+		fedHpa.Status.CurrentReplicas != updatedStatus.CurrentReplicas {
+
+		fedHpa.Status = updatedStatus
+		_, err = a.client.AutoscalingV1().HorizontalPodAutoscalers(fedHpa.Namespace).UpdateStatus(fedHpa)
+		if err != nil {
+			return fmt.Errorf("Error updating hpa: %s status in federation: %v", fedHpa.Name, err)
+		}
+	}
+	return nil
+}
+
+// prepareForScheduling prepares the lists and totals from the
+// existing objs.
+// currentObjs has the list of all clusters, with obj as nil
+// for those clusters which do not have hpa yet.
+func prepareForScheduling(currentObjs map[string]pkgruntime.Object) (hpaLists, replicaNums, map[string]*replicaNums) {
+	lists := hpaLists{
+		availableMax: sets.NewString(),
+		availableMin: sets.NewString(),
+		noHpa:        sets.NewString(),
+	}
+	existingTotal := replicaNums{
+		min: int32(0),
+		max: int32(0),
+	}
+
+	scheduleStatus := make(map[string]*replicaNums)
+	for cluster, obj := range currentObjs {
+		if obj == nil {
+			lists.noHpa.Insert(cluster)
+			scheduleStatus[cluster] = nil
+			continue
+		}
+		if maxReplicasReducible(obj) {
+			lists.availableMax.Insert(cluster)
+		}
+		if minReplicasReducible(obj) {
+			lists.availableMin.Insert(cluster)
+		}
+
+		replicas := replicaNums{min: 0, max: 0}
+		scheduleStatus[cluster] = &replicas
+		if obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas != nil {
+			existingTotal.min += *obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas
+			replicas.min += *obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MinReplicas
+		}
+		existingTotal.max += obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MaxReplicas
+		replicas.max += obj.(*autoscalingv1.HorizontalPodAutoscaler).Spec.MaxReplicas
+	}
+
+	return lists, existingTotal, scheduleStatus
+}
+
+// reduceMinReplicas reduces the min replicas from existing clusters.
+// At the end of the function remainingReplicas.min should be 0 and the MinList
+// and the scheduledReplicas properly updated in place.
+func reduceMinReplicas(remainingMin int32, availableMinList sets.String, scheduled map[string]*replicaNums) int32 {
+	if remainingMin < 0 {
+		// first we try reducing from those clusters which already offer min
+		if availableMinList.Len() > 0 {
+			for _, cluster := range availableMinList.List() {
+				replicas := scheduled[cluster]
+				if (replicas.min - 1) != 0 {
+					replicas.min--
+					availableMinList.Delete(cluster)
+					remainingMin++
+					if remainingMin == 0 {
+						break
+					}
+				}
+			}
+		}
+	}
+
+	// If we could not get needed replicas from already offered min above
+	// we abruptly start removing replicas from some/all clusters.
+	// Here we might make some min to 0 signalling that this hpa might be a
+	// candidate to be removed from this cluster altogether.
+	for remainingMin < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.min - 1) < 0) {
```

Comment:
Realised that I tend to (might however be unsuccessful in many cases) write code in if conditions quite synonymous to stating the if condition. To me atleast it seems more readable. Here it obviously seems to have the undesired effect (you asking, what is it?  :) ). The condition is *if while reducing replica.min by 1, replica.min does not go below 0*. Have updated with *replicas.min > 0* :)